File: panel/src/lib/api/metrics.ts
// The long-term statistics record.
//
// Read-only. There is no write endpoint by design: the value of this record is
// that every number in it was measured, so nothing reachable from a browser can
// put one there.

import { API_URL, getAuthHeader, handleResponse } from '@/lib/api/core';

/** How a series is re-aggregated over a longer window. */
export type MetricKind = 'counter' | 'gauge';

export type MetricUnit = 'count' | 'bytes' | 'bps' | 'percent' | 'seconds';

export interface MetricSeriesInfo {
    metric: string;
    label: string;
    group: string;
    kind: MetricKind;
    unit: MetricUnit;
    help?: string;
    /** Recorded once per component, so a chart has to say whose line it draws. */
    perSubject?: boolean;
}

export interface MetricPoint {
    t: string;
    min: number;
    max: number;
    sum: number;
    count: number;
    /** sum/count - the average of the SAMPLES, not of the buckets. */
    avg: number;
}

export interface MetricSeries {
    metric: string;
    subject?: string;
    points: MetricPoint[];
}

export interface MetricCoverage {
    /** The first sample ever recorded. Null before anything is. */
    since: string | null;
    latest: string | null;
    resolution: string;
}

export interface MetricHeadline {
    metric: string;
    label: string;
    unit: MetricUnit;
    value: number;
    how: 'peak' | 'total' | 'avg';
}

/**
 * Why there is nothing to show, when there is nothing to show.
 *
 * `disabled` is the DEFAULT and not a fault - recording starts when an operator
 * switches it on. `unavailable` means the database could not be reached, which
 * is a problem to fix. Collapsing the two would send someone looking for a
 * broken database when nobody had turned the feature on.
 */
export type MetricsUnavailableReason = 'disabled' | 'unavailable';

/** What every metrics response carries, whether or not it has data. */
export interface MetricsEnvelope {
    available: boolean;
    reason?: MetricsUnavailableReason;
    message?: string;
}

export interface MetricsCatalogResponse extends MetricsEnvelope {
    series?: MetricSeriesInfo[];
    coverage?: MetricCoverage;
}

export interface MetricsSeriesResponse extends MetricsEnvelope {
    series?: MetricSeries[];
    from?: string;
    to?: string;
}

export interface MetricsSummaryResponse extends MetricsEnvelope {
    headlines?: MetricHeadline[];
    coverage?: MetricCoverage;
    from?: string;
    to?: string;
}

/**
 * A network failure is turned into the SAME shape the server sends when there
 * is nothing to show, rather than into the generic {success:false} the other
 * API modules return.
 *
 * The screen then has one thing to render for "no data, and here is why",
 * whether the reason came from the server or from the request never arriving.
 * The alternative is a page that has to distinguish two failure shapes and, in
 * practice, renders blank for one of them.
 */
async function get<T extends MetricsEnvelope>(path: string): Promise<T> {
    try {
        const res = await fetch(`${API_URL}${path}`, { headers: getAuthHeader() });
        return await handleResponse(res);
    } catch (e) {
        console.error('metrics API:', e);
        return {
            available: false,
            reason: 'unavailable',
            message: 'Could not reach the platform.',
        } as T;
    }
}

export const getMetricsCatalog = () =>
    get<MetricsCatalogResponse>('/admin/metrics/catalog');

export const getMetricsSummary = (range: string) =>
    get<MetricsSummaryResponse>(`/admin/metrics/summary?range=${encodeURIComponent(range)}`);

export interface SeriesRequest {
    metrics: string[];
    range: string;
    /** Bucket width in seconds. Omitted lets the server pick a drawable one. */
    step?: number;
    subject?: string;
    region?: string;
    /** One line per component instead of one folded line. */
    split?: boolean;
}

export function seriesQueryString(req: SeriesRequest): string {
    const p = new URLSearchParams();
    p.set('metric', req.metrics.join(','));
    p.set('range', req.range);
    if (req.step) p.set('step', String(req.step));
    if (req.subject) p.set('subject', req.subject);
    if (req.region) p.set('region', req.region);
    if (req.split) p.set('split', '1');
    return p.toString();
}

export const getMetricsSeries = (req: SeriesRequest) =>
    get<MetricsSeriesResponse>(`/admin/metrics/series?${seriesQueryString(req)}`);

/**
 * The export URL.
 *
 * Built rather than fetched: the response is a file download, and the browser
 * has to be the one to receive it. The auth header cannot ride on a plain
 * navigation, so the caller fetches it and hands the blob to a download - see
 * downloadMetricsExport below.
 */
export function metricsExportPath(req: SeriesRequest, format: 'csv' | 'json'): string {
    return `/admin/metrics/export?${seriesQueryString(req)}&format=${format}`;
}

/**
 * Fetch the export and hand it to the browser as a file.
 *
 * Through fetch rather than a link, because the API is behind a bearer token
 * that a plain <a href> cannot carry - a link would arrive unauthenticated and
 * download a 401 body named like a spreadsheet.
 */
export async function downloadMetricsExport(req: SeriesRequest, format: 'csv' | 'json'): Promise<void> {
    const res = await fetch(`${API_URL}${metricsExportPath(req, format)}`, { headers: getAuthHeader() });
    if (!res.ok) {
        throw new Error(`Export failed (${res.status})`);
    }
    const blob = await res.blob();
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = `dylaris-metrics-${new Date().toISOString().slice(0, 10)}.${format}`;
    document.body.appendChild(a);
    a.click();
    a.remove();
    // Released on the next tick rather than immediately: revoking before the
    // browser has started reading the blob cancels the download in Firefox.
    setTimeout(() => URL.revokeObjectURL(url), 1000);
}

/** Formats a value for display, by unit. */
export function formatMetric(value: number, unit: MetricUnit): string {
    switch (unit) {
        case 'bytes':
            return formatScaled(value, ['B', 'KB', 'MB', 'GB', 'TB', 'PB'], 1024);
        case 'bps':
            return formatScaled(value, ['bps', 'Kbps', 'Mbps', 'Gbps', 'Tbps'], 1000);
        case 'percent':
            return `${round(value, 1)}%`;
        case 'seconds':
            return formatDuration(value);
        default:
            return formatCount(value);
    }
}

function round(v: number, places: number): number {
    const f = 10 ** places;
    return Math.round(v * f) / f;
}

function formatScaled(value: number, units: string[], base: number): string {
    let v = Math.abs(value);
    let i = 0;
    while (v >= base && i < units.length - 1) {
        v /= base;
        i++;
    }
    const sign = value < 0 ? '-' : '';
    return `${sign}${round(v, v < 10 ? 2 : v < 100 ? 1 : 0)} ${units[i]}`;
}

function formatCount(value: number): string {
    if (Math.abs(value) >= 1_000_000) return `${round(value / 1_000_000, 1)}M`;
    if (Math.abs(value) >= 10_000) return `${round(value / 1000, 1)}k`;
    // Small counts keep one decimal only when they have one: an average of 3.4
    // players is real, but "12.0 nodes" reads like a bug.
    return Number.isInteger(value) ? String(value) : String(round(value, 1));
}

function formatDuration(seconds: number): string {
    if (seconds < 60) return `${Math.round(seconds)}s`;
    if (seconds < 3600) return `${Math.round(seconds / 60)}m`;
    if (seconds < 86400) return `${round(seconds / 3600, 1)}h`;
    return `${round(seconds / 86400, 1)}d`;
}

/**
 * Which of a point's numbers a series MEANS.
 *
 * A counter is a total (how many handovers happened in this bucket); a gauge is
 * an average (what the CPU was doing across it). Reading a counter as an
 * average understates it by the number of samples in the bucket, and reading a
 * gauge as a total produces a figure with no meaning at all.
 */
export function pointValue(p: MetricPoint, kind: MetricKind): number {
    return kind === 'counter' ? p.sum : p.avg;
}

File: core/handlers/metrics.go
package handlers

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"dylaris-core/metrics"
	"dylaris-core/services"
)

// MetricsHandler serves the long-term record.
//
// Read-only throughout. The recorder is the only writer, and it is leader-gated
// inside Core; there is deliberately no endpoint that can insert a sample, so
// nothing reachable over HTTP can put a number into a record whose whole value
// is that it was measured.
type MetricsHandler struct {
	state *AppState
}

func NewMetricsHandler(state *AppState) *MetricsHandler { return &MetricsHandler{state: state} }

// ready reports whether there is a record to read, and writes the reason if not.
//
// The two "no" cases are different and are told apart on purpose: a metrics
// database that failed to open is a problem to fix, while the feature simply
// being off is the DEFAULT and not a fault. A panel that could not tell them
// apart would send an operator looking for a broken database when nobody had
// switched recording on.
func (h *MetricsHandler) ready(w http.ResponseWriter, r *http.Request) bool {
	if h.state.Metrics == nil || h.state.Metrics.Read == nil {
		sendMetricsJSON(w, map[string]any{
			"success": true, "available": false, "reason": "unavailable",
			"message": "The metrics database is not reachable, so nothing is being recorded.",
		})
		return false
	}
	if !h.state.FeatureFlags.Get(r.Context(), services.MetricsEnabledSetting, false) {
		sendMetricsJSON(w, map[string]any{
			"success": true, "available": false, "reason": "disabled",
			"message": "Long-term statistics are switched off. Turn them on in Settings, Features to start recording.",
		})
		return false
	}
	return true
}

// Catalog GET /api/admin/metrics/catalog
// The series this build records, plus how far back the record goes.
func (h *MetricsHandler) Catalog(w http.ResponseWriter, r *http.Request) {
	if !h.ready(w, r) {
		return
	}
	cov, err := metrics.ReadCoverage(r.Context(), h.state.Metrics.Read, h.state.Metrics.Resolution)
	if err != nil {
		sendJSONError(w, "Could not read the record: "+err.Error(), http.StatusInternalServerError)
		return
	}
	sendMetricsJSON(w, map[string]any{
		"success":   true,
		"available": true,
		"series":    metrics.Catalog,
		"coverage":  cov,
	})
}

// Series GET /api/admin/metrics/series
//
//	?metric=platform.players&from=…&to=…&step=300&subject=…&region=…&split=1
func (h *MetricsHandler) Series(w http.ResponseWriter, r *http.Request) {
	if !h.ready(w, r) {
		return
	}
	q := r.URL.Query()
	from, to, err := windowFrom(q)
	if err != nil {
		sendJSONError(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Several metrics in one request, because a chart page asks for a dozen and
	// a round trip each is the difference between one render and twelve.
	names := splitCSV(q.Get("metric"))
	if len(names) == 0 {
		sendJSONError(w, "Name at least one metric.", http.StatusBadRequest)
		return
	}
	if len(names) > 24 {
		sendJSONError(w, "Too many metrics in one request (24 max).", http.StatusBadRequest)
		return
	}

	step := time.Duration(0)
	if s, err := strconv.Atoi(q.Get("step")); err == nil && s > 0 {
		step = time.Duration(s) * time.Second
	}

	out := make([]metrics.SeriesResult, 0, len(names))
	for _, name := range names {
		res, err := metrics.Query(r.Context(), h.state.Metrics.Read, metrics.SeriesQuery{
			Metric:        name,
			From:          from,
			To:            to,
			Step:          step,
			Subject:       q.Get("subject"),
			Region:        q.Get("region"),
			SplitSubjects: q.Get("split") == "1",
		})
		if err != nil {
			sendJSONError(w, err.Error(), http.StatusBadRequest)
			return
		}
		out = append(out, res...)
	}
	sendMetricsJSON(w, map[string]any{
		"success": true, "available": true, "series": out,
		"from": from.UTC(), "to": to.UTC(),
	})
}

// Summary GET /api/admin/metrics/summary?from=&to=
// The headline numbers, reduced over the window.
func (h *MetricsHandler) Summary(w http.ResponseWriter, r *http.Request) {
	if !h.ready(w, r) {
		return
	}
	from, to, err := windowFrom(r.URL.Query())
	if err != nil {
		sendJSONError(w, err.Error(), http.StatusBadRequest)
		return
	}
	rows, err := metrics.Summary(r.Context(), h.state.Metrics.Read, from, to)
	if err != nil {
		sendJSONError(w, err.Error(), http.StatusInternalServerError)
		return
	}
	cov, _ := metrics.ReadCoverage(r.Context(), h.state.Metrics.Read, h.state.Metrics.Resolution)
	sendMetricsJSON(w, map[string]any{
		"success": true, "available": true, "headlines": rows, "coverage": cov,
		"from": from.UTC(), "to": to.UTC(),
	})
}

// Export GET /api/admin/metrics/export?metric=…&from=&to=&step=&format=csv|json
//
// The point of the export is that the numbers can leave with the person reading
// them - into a spreadsheet, a due-diligence pack, a mail. It is the same query
// the charts run, so an export can never disagree with what was on screen.
func (h *MetricsHandler) Export(w http.ResponseWriter, r *http.Request) {
	if !h.ready(w, r) {
		return
	}
	q := r.URL.Query()
	from, to, err := windowFrom(q)
	if err != nil {
		sendJSONError(w, err.Error(), http.StatusBadRequest)
		return
	}
	names := splitCSV(q.Get("metric"))
	if len(names) == 0 {
		// No metric named means the whole catalog, which is what somebody
		// exporting "the statistics" means.
		for _, s := range metrics.Catalog {
			names = append(names, s.Metric)
		}
	}
	step := time.Duration(0)
	if s, err := strconv.Atoi(q.Get("step")); err == nil && s > 0 {
		step = time.Duration(s) * time.Second
	}

	var all []metrics.SeriesResult
	for _, name := range names {
		res, err := metrics.Query(r.Context(), h.state.Metrics.Read, metrics.SeriesQuery{
			Metric: name, From: from, To: to, Step: step,
			Subject: q.Get("subject"), Region: q.Get("region"),
			SplitSubjects: q.Get("split") == "1",
		})
		if err != nil {
			sendJSONError(w, err.Error(), http.StatusBadRequest)
			return
		}
		all = append(all, res...)
	}

	stamp := time.Now().UTC().Format("2006-01-02")
	if q.Get("format") == "json" {
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="dylaris-metrics-%s.json"`, stamp))
		_ = json.NewEncoder(w).Encode(map[string]any{
			"exported": time.Now().UTC(), "from": from.UTC(), "to": to.UTC(), "series": all,
		})
		return
	}

	w.Header().Set("Content-Type", "text/csv; charset=utf-8")
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="dylaris-metrics-%s.csv"`, stamp))
	cw := csv.NewWriter(w)
	defer cw.Flush()
	// min/max/avg alongside the total, because which of them a series MEANS
	// depends on whether it is a counter or a gauge - and a spreadsheet that
	// only received one column could not recover the others.
	_ = cw.Write([]string{"metric", "subject", "time", "min", "max", "avg", "sum", "count"})
	for _, s := range all {
		for _, p := range s.Points {
			_ = cw.Write([]string{
				s.Metric, s.Subject, p.Time.Format(time.RFC3339),
				strconv.FormatFloat(p.Min, 'f', -1, 64),
				strconv.FormatFloat(p.Max, 'f', -1, 64),
				strconv.FormatFloat(p.Avg, 'f', -1, 64),
				strconv.FormatFloat(p.Sum, 'f', -1, 64),
				strconv.FormatInt(p.Count, 10),
			})
		}
	}
}

// windowFrom reads from/to, as RFC3339 or as a relative "24h"/"30d" in `range`.
func windowFrom(q map[string][]string) (time.Time, time.Time, error) {
	get := func(k string) string {
		if v, ok := q[k]; ok && len(v) > 0 {
			return v[0]
		}
		return ""
	}
	to := time.Now()
	if v := get("to"); v != "" {
		t, err := time.Parse(time.RFC3339, v)
		if err != nil {
			return time.Time{}, time.Time{}, fmt.Errorf("to is not an RFC3339 timestamp")
		}
		to = t
	}
	if v := get("from"); v != "" {
		t, err := time.Parse(time.RFC3339, v)
		if err != nil {
			return time.Time{}, time.Time{}, fmt.Errorf("from is not an RFC3339 timestamp")
		}
		if !t.Before(to) {
			return time.Time{}, time.Time{}, fmt.Errorf("from is not before to")
		}
		return t, to, nil
	}
	// A relative range, which is what the panel's period picker sends. Days are
	// spelled out because "30d" is not something time.ParseDuration accepts.
	span := 24 * time.Hour
	if v := get("range"); v != "" {
		if strings.HasSuffix(v, "d") {
			n, err := strconv.Atoi(strings.TrimSuffix(v, "d"))
			if err != nil || n <= 0 {
				return time.Time{}, time.Time{}, fmt.Errorf("range is not a number of days")
			}
			span = time.Duration(n) * 24 * time.Hour
		} else {
			d, err := time.ParseDuration(v)
			if err != nil || d <= 0 {
				return time.Time{}, time.Time{}, fmt.Errorf("range is not a duration")
			}
			span = d
		}
	}
	return to.Add(-span), to, nil
}

// sendMetricsJSON writes a 200 with a JSON body. Local to this file rather than
// a shared helper: every other handler here builds its own response, and one
// more general utility that only three call sites use is not worth the reach.
func sendMetricsJSON(w http.ResponseWriter, body map[string]any) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(body)
}

func splitCSV(s string) []string {
	var out []string
	for _, p := range strings.Split(s, ",") {
		if p = strings.TrimSpace(p); p != "" {
			out = append(out, p)
		}
	}
	return out
}
